KEP-3570: graduate kubelet cpumanager to GA

fromanirh opened the pull request:
<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- Graduate kubelet CPUManager to GA

<!-- link to the k/enhancements issue -->
- https://github.com/kubernetes/enhancements/issues/3570

<!-- other comments or additional information -->
- Other comments: 
The PR is organized as follows:
1. the first commit is a translation as more mechanical as possible from the ancient document format to the newest available at moment of writing. Care is taken to alter as little as possible the original content, with the only purpose to make it fit in the new format
2. the second commit adds all the new information not required when the original document was written, mostly concern with the PRR reviews.
3. the third commit start updating the content with the plans to GA in 1.26
4. the fourth (and following, if needed) commit adds the new content required by the new format, questionnaires.

fromanirh commented:
/remove-lifecycle stale

fromanirh commented:
trivial rebase

MarSik commented on keps/sig-node/375-cpu-manager/README.md:
```diff
@@ -95,7 +143,34 @@ to build and test new policies. The shared state abstraction allows
+
+#### Story 1 : High-performance applications
+
+Systems such as real-time trading system or 5G CNFs (User Plain Function, UPF) need to maximize the CPU time; CPU pinning ensure exclusive CPU allocation and allows to avoid performance issues due to core switches, cold caches.
```
Isn't it userplane function?

```suggestion
Systems such as real-time trading system or 5G CNFs (User Plane Function, UPF) need to maximize the CPU time; CPU pinning ensure exclusive CPU allocation and allows to avoid performance issues due to core switches, cold caches.
```

fromanirh replied:
it is, it's a typo. I'll fix in the next upload, thanks.

deads2k commented on keps/sig-node/375-cpu-manager/README.md:
```diff
@@ -323,6 +395,240 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+- [ ] Other (treat as last resort)
+  - Details:
+     a operator should check that pods go running correctly and the cpu pinning is performed. The latter can
+     be checked by inspecting the cgroups at node level.
```
Where is this exposed in aggregate?  if it's exposed by something that can be scraped by metrics, that's ok, but we do look for at least one existent implementation of such exposure.

fromanirh replied:
to the best of my knowledge, no well known software component do this, but I'll doublecheck (e.g. node_exporter) and add here as comment for historical record.

deads2k replied:
> to the best of my knowledge, no well known software component do this, but I'll doublecheck (e.g. node_exporter) and add here as comment for historical record.

Being able to know whether a per-node feature is working at scale required for PRR sign off.  Even a simple mis-pinned metric would be sufficient, but ssh-ing to every node is impractical.

Separately, how does a pod author know that his pod is pinned correctly?

fromanirh replied:
Thanks for raising. Point taken, checking. Lack of status reporting can be a gap. Checking and reporting.

mrunalp replied:
cadvisor should have the cpu set information for the pod.

fromanirh replied:
I think we can add two new metrics in `k/k/pkg/kubelet/metrics` to track
1. the total allocation performed by cpu manager which will cause cpu pinning (`cpu_pinning_requests_total`?)
2. the portion of the above which failed (`cpu_pinning_requests_errors_total`?)

this way the cluster admin can have a feedback to check if the cpu pinning feature works and how many allocation failures there are (usually zero), without exposing any unnecessary node-level machine config (e.g. which pinned cores)

could look like this:
```
diff --git a/pkg/kubelet/metrics/metrics.go b/pkg/kubelet/metrics/metrics.go
index 17267b888d7..c8c3a03eaeb 100644
--- a/pkg/kubelet/metrics/metrics.go
+++ b/pkg/kubelet/metrics/metrics.go
@@ -86,6 +86,10 @@ const (
 	// Metrics to track ephemeral container usage by this kubelet
 	ManagedEphemeralContainersKey = "managed_ephemeral_containers"
 
+	// Metrics to track the CPU manager behavior
+	CPUManagerPinningRequestsTotalKey = "cpu_manager_pinning_requests_total"
+	CPUManagerPinningErrorsTotalKey   = "cpu_manager_pinning_errors_total"
+
 	// Values used in metric labels
 	Container          = "container"
 	InitContainer      = "init_container"
@@ -497,6 +501,26 @@ var (
 			StabilityLevel: metrics.ALPHA,
 		},
 	)
+
+	// CPUManagerPinningRequestsTotal tracks the number of times the pod spec will cause the cpu manager to pin cores
+	CPUManagerPinningRequestsTotal = metrics.NewCounter(
+		&metrics.CounterOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           CPUManagerPinningRequestsTotalKey,
+			Help:           "The number of cpu core allocations which required pinning.",
+			StabilityLevel: metrics.ALPHA,
+		},
+	)
+
+	// CPUManagerPinningErrorsTotal tracks the number of times the pod spec required the cpu manager to pin cores, but the allocation failed
+	CPUManagerPinningErrorssTotal = metrics.NewCounter(
+		&metrics.CounterOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           CPUManagerPinningErrorssTotalKey,
+			Help:           "The number of cpu core allocations which required pinning failed.",
+			StabilityLevel: metrics.ALPHA,
+		},
+	)
 )
 
 var registerMetrics sync.Once
```
(names and descriptions subjected to change). I'll file a separate demo PR and update this one ASAP, since I think this is a nice monitoring improvement anyway.

fromanirh replied:
implementation sketch: https://github.com/kubernetes/kubernetes/pull/112855

swatisehgal replied:
Let's capture these proposed metrics in the KEP. If @derekwaynecarr and @deads2k are okay with it we can get https://github.com/kubernetes/kubernetes/pull/112855 merged before 1.26 code freeze. 
We can gather feedback from other reviewers regarding the metrics to be gathered (and naming) as part of the KEP review. I am happy with the approach so far.

fromanirh replied:
> Let's capture these proposed metrics in the KEP. If @derekwaynecarr and @deads2k are okay with it we can get [kubernetes/kubernetes#112855](https://github.com/kubernetes/kubernetes/pull/112855) merged before 1.26 code freeze. We can gather feedback from other reviewers regarding the metrics to be gathered (and naming) as part of the KEP review. I am happy with the approach so far.

sure thing. I'll update the KEP proper ASAP.

deads2k replied:
The metrics described in this thread look good to me.  If you update the KEP with these names (even before the impl merges), that will do.

fromanirh replied:
thanks, on it.

deads2k commented on keps/prod-readiness/sig-node/375.yaml:
```diff
@@ -0,0 +1,7 @@
+kep-number: 375
```
this looks to be linked to https://github.com/kubernetes/enhancements/issues/3570 for kube enhancements tracking

fromanirh replied:
Done. I updated the KEP number with the newest one, because it's a bit more practical to work on. Added backlinks (and used `git mv`) to preserve the history and have clear history. I think this should be good enough for the project purposes; if this is not the case I'd be most happy to change direction.

fromanirh commented:
/retitle KEP-375: graduate kubelet cpumanager to GA

fromanirh commented:
removed WIP. Will handle the KEP number mismatch ASAP

swatisehgal commented on keps/sig-node/3570-cpumanager/README.md:
````diff
@@ -156,11 +228,12 @@ type CPUSet map[int]struct{} // set operations and parsing/formatting helpers
 ```
 
-#### Configuring the CPU Manager
+### Configuring the CPU Manager
````
As `CPUManagerPolicyOptions` allows us to tweak the behavior of static policy, we should probably add some information about it here. Just providing a list of policy options and adding references to the relevant KEPs should be sufficient:

1.  https://github.com/fromanirh/enhancements/tree/master/keps/sig-node/2625-cpumanager-policies-thread-placement
2. https://github.com/fromanirh/enhancements/tree/master/keps/sig-node/2902-cpumanager-distribute-cpus-policy-option
3. https://github.com/fromanirh/enhancements/tree/master/keps/sig-node/3327-align-by-socket

fromanirh replied:
good point, done.

swatisehgal commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +396,240 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [ ] Feature gate (also fill in values in `kep.yaml`)
```
Adding a comment so we don't forget to fill this section.

fromanirh replied:
Good point, thanks. Fixed.

swatisehgal commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -1,91 +1,140 @@
 
-_Solution requirements:_
+### Goals
 
 1. Provide an API-driven contract from the system to a user: "if you are a
    Guaranteed pod with 1 or more cores of cpu, the system will try to make
    sure that the pod gets its cpu quota primarily from reserved core(s),
    resulting in fewer context switches and higher cache affinity".
-1. Support the case where in a given pod, one container is latency-critical
+2. Support the case where in a given pod, one container is latency-critical
    and another is not (e.g. auxiliary side-car containers responsible for
    log forwarding, metrics collection and the like.)
-1. Do not cap CPU quota for guaranteed containers that are granted
+3. Do not cap CPU quota for guaranteed containers that are granted
    exclusive cores, since that would be antithetical to (1) above.
-1. Take physical processor topology into account in the CPU affinity policy.
-
-### Related issues
+4. Take physical processor topology into account in the CPU affinity policy.
```
A NIT which was pointed out by SIG docs in one of my PRs: for ordered list we should use `1` for all the items as the markdown parser captures the correct number. This is beneficial if items are added or removed requiring no updates to the numbering.

fromanirh replied:
TIL! thanks, fixing (all across the document)

swatisehgal commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +396,240 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
```
Looks like a comment trailing tag, needs to be removed.

fromanirh replied:
Indeed. Thanks, fixed.

swatisehgal reviewed (commented):
Looks good overall, left a few minor comments.

deads2k commented on keps/sig-node/375-cpu-manager/README.md:
```diff
@@ -323,6 +395,240 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+###### What specific metrics should inform a rollback?
+
+Pod creation errors o a node-by-node basis.
```
typo and a specific metric name is needed here.

fromanirh replied:
typo fixed, understood about the metric and working on it. Can be a gap to address.

deads2k commented on keps/sig-node/375-cpu-manager/README.md:
```diff
@@ -323,6 +395,240 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+###### How can an operator determine if the feature is in use by workloads?
+
+The operator need to inspect the node and verify the cpu pinning assignment either checking the cgroups on the node
```
CPU pinning is indicated on the pod somewhere, right?  Which field on the pod indicates usage?

fromanirh replied:
the request of cpu pinning is requested by the pod

1. opting in the "guaranteed" QoS
2. requesting integral number of CPUs

there is no single value (e.g. field in pod status) which however report if pinning was succesfull. Working on it. *Can* be a gap, will check and report.

derekwaynecarr replied:
@deads2k cpu pinning is a node local decision.

if the node is configured to support the feature (which is not advertised to the scheduler because its a node local optimization), and the pod is in the guaranteed qos class and advertises integral cpus, the kubelet if configured will pin the pod to dedicated cpus.

there is no 'pin this pods cpu' feature advertised in the pod API itself by design.

fromanirh replied:
The resources request: `pod.spec.containers[*].resources.limits.cpu`. If it's integral value (e.g. 2) and if the pod QoS is guaranteed, this combination will trigger the exclusive CPU allocation and the pinning (https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy)

fromanirh commented on keps/sig-node/375-cpu-manager/README.md:
```diff
@@ -323,6 +395,240 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+### Monitoring Requirements
+
+Monitor the pod admission counter
```
ack, working on it

fromanirh commented:
addressed most of review comments, but not those about the lack of metrics. This is understood and still being worked on. Can be a gap to address before to GA.

derekwaynecarr commented:
@deads2k The admin can inspect this feature is functional by seeing the assigned cpuset for the pod via existing monitoring tools that are cgroup aware, so I am not sure a metric is truly needed.  Mispinned is not a concept either, if the pod won't fit, the pod is rejected on admission with an appropriate failure condition.  There is no in-between state.

@fromanirh possibly adding a metric for the number of pods this kubelet failed at admission due to unavailable cpu set capacity would be useful, but I am not sure I see folks in practice looking at that as much as looking at the pod failure condition.

andrewsykim commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +407,236 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+No, because all the metrics we were aware of leaked hardware details.
```
Are there are Pod metrics around # of Guarenteed pods? It seems like a metric could be derived from checking if a Pod is Guaranteed and has an integer CPU request/limit?

andrewsykim replied:
Oh, I see this was already brought up here https://github.com/kubernetes/enhancements/pull/3350#discussion_r985983166

fromanirh replied:
yes, I elaborated this point after further discussion, in the same direction you mentioned.

andrewsykim commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +407,236 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+No
```
Seems useful to describe how resource usage could change in a give node when enabling this feature. Even though total amount of CPU time is the same, pinning CPUs would change resource usage is a non-neglible way right?

fromanirh replied:
That's a good point. I'll think and try to elaborate. Perhaps a third metric can help here.

andrewsykim commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +407,236 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+###### What are other known failure modes?
+
+After changing the CPU manager policy from `none` to `static` or the the other way around, before to start the kubelet again,
+you must remove the CPU manager state file(`/var/lib/kubelet/cpu_manager_state`), otherwise the kubelet start will fail.
```
Just out of curiosity, does kubelet know about this state and make a recommendation to users to remove this state file?

fromanirh replied:
The flow here is probably not great. Changing the policy and restarting the kubelet will very likely (= almost always in my experience) lead the checkpoint recovery code fail, so we will end up in this flow: https://github.com/kubernetes/kubernetes/blob/v1.25.2/pkg/kubelet/cm/cpumanager/cpu_manager.go#L223

So I guess the answer is "yes", users are notified by the kubelet going in a crashloop/refusing to start, with this line in the logs.

andrewsykim replied:
Gotcha -- for GA we should consider improving that log line. Looking at it on first glance it's not immediately clear which checkpoint manager the log line refers to and which policy state file this is for. Even something like:
```
klog.ErrorS(err, "Could not initialize CPU Manager checkpoint manager, please drain node and remove policy state file %q", cpuManagerStateFileName)
```

fromanirh replied:
Very fair point. I think we can totally do this before GA'ing the feature.

andrewsykim replied:
> EDIT - the only caveat is that if my memory serves, the file name (which will be cpu_manager.go) is logged so it can be somehow inferred by the context. Anyway I believe a error message tuning is no bother.

Yeah good point -- I think the real change needed is just logging the state file path so it's immediately clear which file needs to be removed

andrewsykim commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -156,11 +229,12 @@ type CPUSet map[int]struct{} // set operations and parsing/formatting helpers
-Kubernetes will ship with three CPU manager policies. Only one policy is
+Kubernetes will ship with CPU manager policies. Only one policy is
 active at a time on a given node, chosen by the operator via Kubelet
-configuration. The three policies are **none**, **static** and **dynamic**.
```
Out of curisoity -- why was the `dynamic` policy removed?

fromanirh replied:
I don't have this information. I started contributing in this area way after the original discussion. I can *guess* the scope was downsized, and later the use case for this dynamic policy faded out.

andrewsykim commented on keps/sig-node/3570-cpumanager/kep.yaml:
```diff
@@ -0,0 +1,50 @@
+# The most recent milestone for which work toward delivery of this KEP has been
+# done. This can be the current (upcoming) milestone, if it is being actively
+# worked on.
+latest-milestone: "v1.10"
```
This should be `v1.26`

fromanirh replied:
thanks, fixed

fromanirh replied:
thanks, fixed

andrewsykim commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +407,236 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+No
```
Oh, I see this was already brought up here https://github.com/kubernetes/enhancements/pull/3350#discussion_r985983166

fromanirh replied:
yes, I was too terse :) Elaborating like you suggested.

deads2k commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +407,245 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+"cpu_manager_pinning_requests_total" and "cpu_manager_pinning_errors_total"
```
This won't be missing when GA, right? Can you move this into the "What are the reasonable SLOs (Service Level Objectives) for the enhancement"?  We will trust you to add them

fromanirh replied:
I'm 100% committed in having them in: [issue](https://github.com/kubernetes/kubernetes/issues/112854) and [PR](https://github.com/kubernetes/kubernetes/pull/112855) are up to already. From my PoV the only thing missing in the PR is some form of testing to ensure the metrics are populated.

deads2k commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +407,245 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+###### How can an operator determine if the feature is in use by workloads?
+
+The operator need to inspect the node and verify the cpu pinning assignment either checking the cgroups on the node
```
Can you describe how to inspect pods to know if the field is set properly, you have a comment that describes how to do this in the PR.

fromanirh replied:
done.

deads2k commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +407,245 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+
+- [X] Other (treat as last resort)
+  - Details: the containers need to check the cpu set they are allowed to run; in addition, node agents (e.g. node_exporter)
```
with your metrics update, they know this based on the `cpu_manager_pinning_requests_total` metric.

deads2k commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +407,245 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+### Monitoring Requirements
+
+Monitor the pod admission counter
```
is there a kubelet metric for this?  Will you add one in order to go GA?

deads2k commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +407,245 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+### Monitoring Requirements
+
+Monitor the pod admission counter
+Monitor the pods not going running after successful schedule
```
is there a metric for this already exposed?

fromanirh replied:
Since exclusive cpu allocation is done as admission time, we can just check the admission result. This is already being reflected in the pod status.

swatisehgal commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -323,6 +407,257 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+but adding these two metrics is both very cheap and helping for the observability of the system.
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
```
We should mention the metrics here instead of selecting `Other`:
```
- [X] Metrics
  - Metric name:
    - cpu_manager_pinning_requests_total
    - cpu_manager_pinning_errors_total
```

fromanirh replied:
thanks, changing.

swatisehgal reviewed (commented):
Looking good, left a couple of minor comments.

fromanirh commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -95,7 +145,30 @@ to build and test new policies. The shared state abstraction allows
+
+### Risks and Mitigations
+
+TBD
```
THanks, worth elaborating indeed.

deads2k commented:
PRR looks good. Thank you for the updates.

/approve

swatisehgal commented:
Thanks for promptly addressing all the review comments.

/lgtm 
from node-perspective (non-binding)

swatisehgal commented:
/cc @derekwaynecarr @dchen1107 @mrunalp @klueska 
for sig-node approval

mrunalp commented:
/approve

mrunalp commented:
/assign @dchen1107

dchen1107 commented on keps/sig-node/3570-cpumanager/README.md:
```diff
@@ -1,43 +1,113 @@
+- [X] "Implementation History" section is up-to-date for milestone
+- [X] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
```
The GA criteria list looks good to me.

dchen1107 commented:
/lgtm
/approve
